### ISSUE TITLE
Add support for multi-valued tags

### ISSUE THREAD

**Issue by ryanakca** [beetbox/beets#505]
Please add support for multi-valued tags. For example, this could be used for the multi-valued occurrences of PERFORMER / COMPOSER / ENSEMBLE / etc. recommended by http://age.hobba.nl/audio/mirroredpages/ogg-tagging.html .

**Comment by sampsyo** [comment]
Good summary; thanks!

The reason we haven't done 4 is because it will also make queries more complicated. Currently, there is a uniform way to construct queries for any fixed field. Having a separate table for the list-valued fields would mean that querying would look very different depending on the type of the field. It's also not 100% clear how this would work with flexible attributes, or how flexible attributes could be migrated to fixed attributes.

In summary: it could definitely be the best idea, but someone needs to do a detailed design to show exactly how this would work.

**Comment by pdf** [comment]
I looked at extending some of the beets code some years ago now (though I don't normally write Python), but I gave up when I started looking at the DB stuff.  Making the entire database layer bespoke doesn't seem labor-efficient, or: _How I Learned to Stop Worrying and Love the ORM_.

The kludge in 3 would probably solve most people's desire here, and it's pretty trivial by comparison.

**Comment by sampsyo** [comment]
As you noted, 3 has not been on the table mostly because querying is too weird. Queries are a core concept.

**Comment by pdf** [comment]
That's only a problem if you care about querying these relationships - for me, I just want my tags right, without having to use a GUI (ie - Picard).  Other tickets have been created for the same reason.  I understand the desire to architect it elegantly though.

**Comment by Boehemyth** [comment]
One thing I had started looking into (before I got a new job and got too busy) was moving beets over to ZODB instead of sqlite.  This would basically be a core overhaul, but it would really make sense in the long run.  ZODB is a object oriented database that would basically allow us to keep track of lists inside our records (and query them).  The problem is this will take a great deal of effort, but I think it is better than moving to a document oriented database like we've been talking about here.

**Comment by sampsyo** [comment]
I looked at ZODB as an alternative too, and it has lots of great features. I was a little turned off by tapping into Zope ecosystem, for fear of too many dependencies and perceived lock-in—it would definitely be nice to be able to interoperate with non-Python tools. Still worth considering though.

**Comment by Boehemyth** [comment]
>  I was a little turned off by tapping into Zope ecosystem, for fear of too many dependencies and perceived lock-in

A big disadvantage of ZODB over SQLite is that it would require extra dependencies since it isn't built into Python like SQLite is.  The other thing that I wasn't sure about was whether the Zope Public License, which ZODB is licensed under, is compatible with the MIT License.

**Comment by sampsyo** [comment]
I'm less worried about licensing, since the ordinary Python distribution workflow ships libraries as separate components.

**Comment by GuilhermeHideki** [comment]
Is PostgreSQL with standard types, ``key:val`` and/or ``JSON`` an option?

**Comment by Rouzax** [comment]
Just curious what the status of this is, because it relates directly with my issue of multiple genres #1845

**Comment by perfectcolors** [comment]
I'm interested in this too as it's one of the issues keeping me from having beets manage my music library.

Right now I just use beets to pre-process / import my music to a temporary folder (it's great for this and still saves a lot of time) then I use a separate tag editor to create multi-value entries for the artist field and manually copy the files into my main library from there.

**Comment by m-urban** [comment]
I haven't followed the entire discussion, but I wonder If we are making this too complicated, trying to come up with the *perfect* solution.
One suggestion was to serialize lists to a string with a unique separator. It was mentioned that this would introduce a lot of overhead because such strings would need to be de-serialized in each query. While this might generally be true, I think that the actual problem is a lot less complex.

Which tags could particularly benefit from the list feature? I think, it is mostly string values, which aren't any more difficult to query when containing a separator for multiple values. I am thinking artists, composers, performers, lyricists, conductors, genres, etc. It was stated before that numbers would be problematic. Maybe I am just lacking imagination, but from the "standard tags", I cannot think of an integer value where a list would make sense. Year? Track? Disc?

**Comment by sampsyo** [comment]
That's a good point—it might be worth at least prototyping a delimiter-based solution that doesn't alter the database schema. A couple of questions worth investigating first:

* Does SQLite support strings with null bytes in them?
* How much of a performance hit is there when querying by delimited substring vs. by equality?

**Comment by tjconnelly** [comment]
fwiw this is the feature blocking my total adoption as well. i'm working on a playlist manager, so i can hack a workaround using playlists as tags in the interim, but meanwhile:

general +1, and furthermore a vote to make the tags queryable.

i'm just getting my feet wet, but could concerns around 4 above be handled in the configuration? ie, could there be a setting added to customize which attrs are available for search in a given library and construct from there?

thx for eyeballs and code; i love that beets exists.

**Comment by sampsyo** [comment]
Maybe so! That comes with its own set of strange decisions, of course: what happens when you change a field from single-valued to multi-valued? What if a tag you set as multi-value can only be single-valued in the underlying tag format?

Overall, the ideal solution (for me) would treat all tags as potentially multi-valued, but under most circumstances you could treat them as single-valued with no additional complexity—you'd just be setting and retrieving the 0th element of the value list. Alternative querying and update mechanisms would let you access the whole list.

**Comment by antlarr** [comment]
> Does SQLite support strings with null bytes in them?

Just for the records, I did a quick test and it seems it's supported:

```
>>> import sqlite3
>>> conn = sqlite3.connect('test.sqlite')
>>> c = conn.cursor()
>>> c.execute('create table songs (id INTEGER PRIMARY KEY ASC AUTOINCREMENT, genre TEXT );')
<sqlite3.Cursor object at 0x7f4949ad5260>
>>> g=u'Rock\0Blues'
>>> c.execute('INSERT INTO songs(genre) VALUES (?)', (g,))
<sqlite3.Cursor object at 0x7f4949ad5260>
>>> b=c.execute('SELECT genre from songs')
>>> b.fetchall()
[('Rock\x00Blues',)]
```
The problem is I don't think this can't be used nicely to query the database. Also, if you run the sqlite3 command line interface, it seems not to be supported:

```
> sqlite3 test.sqlite 
SQLite version 3.17.0 2017-02-13 16:02:40
Enter ".help" for usage hints.
sqlite> select genre from songs;
Rock
```

**Comment by pprkut** [comment]
Rather than strings with a specific separation character, maybe storing JSON arrays/objects could be an option.
Sqlite does have native support for working with JSON now, although it's probably not very widespread yet: https://sqlite.org/json1.html

**Comment by antlarr** [comment]
That's quite interesting. I didn't know that sqlite extension. Right now the options would be:
- Store the list joining the elements with a separator. This is what's currently implemented in #2503 .
- Convert to json array and store that. That could be easily implemented, and would allow (not so easily, and considering I didn't get it wrong) to do complex queries using the json1 extension you linked.
- Use a separate table (could the flexible attributes table be used for that? or should another table be created for this?) and remove the column from items/albums tables. In my opinion, this would allow to use simpler code in the future.

What do you think? I think I would prefer the third option (the normalized database schema) better, but I'm open to other opinions.

**Comment by pprkut** [comment]
I think starting off with a simple implementation using a separation character is fine, if there is a forward-migration path (i.e. we could switch to a different implementation later).
That implementation is already (almost) there and would make a lot of users perfectly happy, so it makes sense to consider that.

I mentioned JSON because it's still readable as raw value in the database and slightly more structured than a separation character. It would also open up opportunities to use JSON values for other more complex fields, like for example storing the raw values from acousticbrainz. The downside of JSON is that it's slightly less useful as is in paths, so we'd need to think about how to normalize it for that.

Separate tables would be great for organizing the data properly, but carry significant performance drawbacks when querying information from the database. We see this already with the flexible attributes. The problem is that values stored in the items table in a single cell allow for a single result set for multiple items. A separate, flexible attribute like table would require one query/result set per item. If you query your library for 5000 tracks, that's a difference of 5000 queries. (There's probably ways to cut down the amount of queries, but that would likely result in larger code changes).

**Comment by sampsyo** [comment]
I suddenly like the idea of JSON as a near-term solution—it might be less of a headache than null separators. In particular, the path formatting thing isn't exactly a problem: we already use custom formatting routines to convert internal values into strings before formatting templates. So users should not be able to notice the internal storage format.

Of course, it would make querying a bit harder unless we have a SQLite with JSON support.

**Comment by antlarr** [comment]
At least in openSUSE, sqlite has json support by default. Do you know if that's a reasonable dependency that can be added to beets? If so, I'll try to use change my PR to use json values. BTW, would it make sense to add a database scheme versioning? I'm thinking it would be interesting if we don't mix regular strings values with json values in a given column, so some code to "migrate" the column values would be needed (that is, in database format v1, genre, etc store strings, while in v2, json values are stored).

Just for the records:
```
sqlite> CREATE TABLE songs (id INTEGER PRIMARY KEY ASC AUTOINCREMENT, genre TEXT );
sqlite> insert into songs (genre) VALUES (json_array('rock', 'blues'));
sqlite> insert into songs (genre) VALUES (json_array('pop'));
sqlite> select songs.id from songs, json_each(songs.genre) as json_genre where json_genre.value = 'rock';
1
sqlite> select songs.id from songs, json_each(songs.genre) as json_genre where json_genre.value like '%o%';
2
3
```
So this looks promising :)

**Comment by sampsyo** [comment]
Good question—I don't know how pervasive that version of SQLite is. That would require some careful searching into the state of various OSes. On the other hand, nothing stops us from storing JSON in our database even if we can't use the fancy queries over it yet.

I suppose you're right that the nice thing about the null-separated encoding is that there is no migration required at all. A single-element list is exactly the same as a plain string.

**Comment by sampsyo** [comment]
About versioning: so far, we've been able to avoid it. The schema benefits from being so incredibly simple that we never really need to grapple with complicated migrations. We might need to explore it now, of course!

**Comment by mafrasi2** [comment]
> I don't know how pervasive that version of SQLite is. That would require some careful searching into the state of various OSes

As I'm very interested in this feature, I did some research on this:

    Debian (and probably Ubuntu): enabled in stretch since 2015-10-19
                                  disabled in jessie and older
    Arch    : enabled since 2016-01-19
    Fedora  : enabled since 2016-01-22
    CentOS  : disabled
    OpenSUSE: enabĺed
    FreeBSD : disabled by default, but simple to enable
    Windows : enabled
    Mac OS X: disabled according to https://docs.python.org/3/library/sqlite3.html#f1

The biggest barrier is probably Mac OS X.

**Comment by sampsyo** [comment]
Hmm; that is tricky. As usual, macOS is behind.

Maybe we can architect this feature so it degrades gracefully on systems without the appropriate SQLite feature?

**Comment by divanikus** [comment]
While you guys thinking of a some generic solution I came up with the following hack of mediafile.py
```python
class ListMediaField(MediaField):
    """Property descriptor that retrieves a list of multiple values from
    a tag.

    Uses ``get_list`` and set_list`` methods of its ``StorageStyle``
    strategies to do the actual work.
    """
    def __get__(self, mediafile, _):
        values = []
        for style in self.styles(mediafile.mgfile):
            values.extend(style.get_list(mediafile.mgfile))
        return [_safe_cast(self.out_type, value) for value in values]

    def __set__(self, mediafile, values):
        for style in self.styles(mediafile.mgfile):
            style.set_list(mediafile.mgfile, values)

    def single_field(self):
        """Returns a ``MediaField`` descriptor that gets and sets the
        first item.
        """
        options = {'out_type': self.out_type}
        return ListMediaSingleField(*self._styles, **options)



class ListMediaSingleField(MediaField):
    """Property descriptor that retrieves a list of multiple values from
    a tag.

    Uses ``get_list`` and set_list`` methods of its ``StorageStyle``
    strategies to do the actual work.
    """
    def __get__(self, mediafile, _):
        values = []
        for style in self.styles(mediafile.mgfile):
            values.extend(style.get_list(mediafile.mgfile))
        return "; ".join([_safe_cast(self.out_type, value) for value in values])

    def __set__(self, mediafile, value):
        values = value.split("; ")
        for style in self.styles(mediafile.mgfile):
            style.set_list(mediafile.mgfile, values)
```
Combined with 
```yaml
lastgenre:
  canonical: yes
  count: 5
  separator: '; '
```
Works pretty good to me. At least for mp3s and flacs. 
I know that's not a production quality solution, but it just works in my case: beets can both read and write multiple genre tags.

**Comment by GuilhermeHideki** [comment]
I was tinkering with the join table for flexible attributes and lists (i guess this is kind of horizontal partitioning). I don't know how much "magic" it would need for creating the query from user input +  "metadata" table, like in CMSs which you can customize the fields. The idea would be like this:

```SQL
CREATE TABLE songs (
  id INTEGER PRIMARY KEY AUTOINCREMENT, 
  title TEXT
  --   ...default fields
);

-- list type
CREATE TABLE songs_artists (
  id INTEGER, 
  ord INT DEFAULT 1, 
  name TEXT NOT NULL, 
  joiner TEXT, -- feat
  PRIMARY KEY (id, ord),
  FOREIGN KEY (id) REFERENCES songs(id), 
)
```

```py
# example
fields = OrderedDict({
  'title': 'songs.title',
  'artists': 'GROUP_CONCAT(songs_artists.joiner || songs_artists.name, "")'
})

joins = [
  'JOIN songs_artists on songs.id = songs_artists.id
]

filter_subselect = f"SELECT id from songs_artists where {criteria}"
sql = f"SELECT {','.join(fields.values())} FROM songs {joins} WHERE songs.id IN (filter_subselect)"

# fetch and create objects / DTOs etc
```

**Comment by sampsyo** [comment]
Sure! This seems like about the right design to make arbitrary files into lists. I'd be interested to investigate what the potential performance trade-offs are…

**Comment by Freso** [comment]
> Maybe we can architect this feature so it degrades gracefully on systems without the appropriate SQLite feature?

For PostgreSQL I've seen people recommend converting dbs with JSON fields to regular string fields and rely on downstream software to decode the JSON when dealing with pgsql versions prior to native JSON support. Maybe beets could also just use [`json`](https://docs.python.org/3/library/json.html#module-json) if we can't get/store "real" JSON objects from native SQLite?

**Comment by jackwilsdon** [comment]
@Freso that'd mean we'd need two versions of our queries to degrade nicely for versions of SQLite without `json1` support, as the main advantage to using the JSON column type are the functions supported by the SQLite `json1` extension.

**Comment by Freso** [comment]
@jackwilsdon I think there are some other great advantages, e.g., that you can store lists of non-string objects (incl. other lists!), which is not possible with the "CSV" approach (regardless of what we decide our "C" to be). (We haven't found a use case for these though, but it's nice to go down a path that is known to be easily extendable.)

I do agree that it's a bother, and I wouldn't mind if it was decided to be a beets 2.0.0 release feature that simply has a hard requirement for `json1` SQLite support. Older distributions and builds that don't have it enabled would then have to use beets <2, which is also working fine for most cases. (Mostly I was just commenting to hopefully get the issue moving along a bit. (Four years ago, @sampsyo said "We should indeed get moving on list-valued fields."…) :))

**Comment by pdf** [comment]
What's wrong with just using the standard relational DB option of a join table? Losing indexing seems like a poor trade-off for storing simple tabular data as JSON.

**Comment by jackwilsdon** [comment]
@Freso I'm certainly not against storing it as JSON, my point was mainly around we the fact that we wouldn't be able to use any advantages of `json1` until we deprecated support for non-`json1` versions of SQLite, which is less than ideal.

I guess one thing to look at is % of installations that support `json1` — we'd probably have to go digging in different distros or maybe outsource some of the statistics to the community if they can't be found elsewhere.

**Comment by jtpavlock** [comment]
Fwiw, [starting with 3.9, json1 (sqlite extension) will be included by default in python](https://bugs.python.org/issue40270). 

I'm not able to comment on the pros and cons of using json vs join tables, but I am hoping the discussion continues so we can begin working on something :)

**Comment by sampsyo** [comment]
That's great news; thanks for the heads up!

**Comment by jtpavlock** [comment]
Could something like this python package, [supersqlite](https://github.com/plasticityai/supersqlite), be used to bridge the gap until the features we need for json support are officially added?

**Comment by Vesnyx** [comment]
Since python 3.9.3 is now officially released, it should have json1 support (https://github.com/python/cpython/commit/58d6f2ee3aeb699156d4784acccd2910d27982e7). Has anybody begun working on migrating this? I suspect that it will be an arduous task. If not, I might try taking a stab at it.

@beetbox Will we need to include backwards support for older databases or would it be better to have a forced database upgrade? I would love to use jsons to store lists and there are quite a few other features that could benefit as a result of this upgrade.

**Comment by wisp3rwind** [comment]
> Since python 3.9.3 is now officially released, it should have json1 support ([python/cpython@58d6f2e](https://github.com/python/cpython/commit/58d6f2ee3aeb699156d4784acccd2910d27982e7)). Has anybody begun working on migrating this? I suspect that it will be an arduous task. If not, I might try taking a stab at it.
> 
> @beetbox Will we need to include backwards support for older databases or would it be better to have a forced database upgrade? I would love to use jsons to store lists and there are quite a few other features that could benefit as a result of this upgrade.

Personally, I don't see an issue with forcing a database upgrade. That's probably simpler to implement and more robust than attempting to be backwards compatible? AFAIK, up to now, beets doesn't have any serious database upgrade code (except for adding new columns to the album and item tables), but I don't see a problem with adding it. We may consider a commandline prompt where users need to acknowledge the irreversible upgrade and which we could use to encourage backing up the database first.

I'm not so sure that we want to require Python 3.9.1 (released in August 2020) just yet, that's still somewhat recent. I don't think we have any idea what Python versions are common among beets users, unfortunately. I wouldn't say that we need to retain support for any Ubuntu LTS that's still out there. But maybe at least 3.8? That's not to say that work on this feature can't be started now, but I guess a few more opinions would be nice to set a realistic timeline for actually shipping json1.

**Comment by animaldaydream** [comment]
I don't want to be dismissive, but I just wanna say: What would the venn diagram of "Ubuntu users" + "beets users" look like? I bet there's not much overlap. Now if we were to say: "Ubuntu users" + "Picard users" maybe there's quite a bit of overlap.

Either way, those who are running Ubuntu (unless they use pip) are unlikely to upgrade to a new version of beets until a new release of Ubuntu appears, at which point the package gets upgraded on the Ubuntu repos.

So if you want to serve those using Ubuntu, but not pip, you'd fix up a database transition from latest Ubuntu/Ubuntu LTS to the next version of beets.

And if you want to serve those using pip on Ubuntu, perhaps it's better to just ask them to upgrade the relevant packages. They _did_ choose to install from pip. Ubuntu users eventually learn that using Ubuntu means a commitment to running the version of the packages specifically packaged for their version of Ubuntu. If you don't, you can mess up the system stability. Truth is, it's a big commitment to run latest packages on Ubuntu.

Which brings me to my next point: Why not let us choose when to migrate the database? Perhaps you can leave a setting to use the single-value database until we can migrate. I don't know how big of a job it is to leave legacy support though. You have to figure that out yourselves.

That's my two cents as user and not developer.

**Comment by jtpavlock** [comment]
Another option could be to require different python versions per OS. MacOS should support json1 [starting with python 3.7](https://github.com/python/cpython/commit/9625bf520e08828e36bc3b1d043af679eb5f993d), while I know most linux OSs have been compiling their sqlite with json1 for a while ([Ubuntu since at least 16.04](https://stackoverflow.com/questions/49043084/sqlite-on-ubuntu-16-04-with-json1)).

So really the python requirement would be 3.7 on MacOS, 3.9 on Windows, and a somewhat recent linux distribution (my understanding is that json1 on linux is an OS-specific thing rather than a python specific thing)

**Comment by Freso** [comment]
> Why not let us choose when to migrate the database?

You have that option by choosing when you update beets. If you’re not ready to take the jump, just don’t update beets with the new db schema until you are. :shrug:

The massive overhaul this will be to how beets works internally and with its database would make it extremely clunky to support the two different schemas simultaneously. It’s already a task that we’ve been _wanting_ to do/implement ASAP for 7+ years, requiring backwards compatibility with old databases would only add to the already massive task of accomplishing this.

**Comment by MountainX** [comment]
What's the latest status update on this?

**Comment by kifty** [comment]
This is the one feature of Beets that's keeping me from going "all in". I have thousands of files that are tagged with multiple genres.

**Comment by jmbannon** [comment]
I'm attempting to add this for `artists`, `albumartists`, and `genre`.

This is my first time ever dabbling in the beets codebase, so I probably won't know what I'm doing for a while, lol. But nonetheless, I'm determined to make this work. 

Perhaps @jpluscplusm  could share some wisdom regarding the albumtypes addition from a few weeks ago? I'm thinking the above tags could work similarly

**Comment by jmbannon** [comment]
After carefully reading through the prior comments, I am wondering why json seems to be preferred over using a simple null-valued separator in a string field? I think the performance impact would be minimal, and require the least headache in terms of dependencies, migration, etc. I would also think making non-list fields into json for the sake of 'fields being agnostic' would incur more compute cost.

At this point, the issue has been open for almost 7 years. I think most users who care about this issue will take any solution.

I will be attempting the 'null separator in a string field' solution to this problem.

**Comment by jmbannon** [comment]
@JOJ0 I think we can close this (right before the 10 year mark 😅)

**Comment by NicolasPL64** [comment]
Sorry for bringing this up again, but I can't find any documentation on how to get multiple artists and multiple album artists working with Beets. All I get is them joined with the corresponding MusicBrainz join character. I'd love having an option to replace it with `; ` or `\\`.

**Comment by snejus** [comment]
@NicolasPL64 which beets version are you using? When you run `beet list -f $artists` they should be automatically joined by `; `.
